cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,356 @@
+template <typename T>
+__global__ void compute_percentiles_kernel(offset_type const* tdigest_offsets,
+                                           size_type num_tdigests,
+                                           device_span<double const> percentiles,
+                                           double const* mean_,
+                                           double const* weight_,
+                                           double const* min_,
+                                           double const* max_,
+                                           double const* cumulative_weight_,
+                                           device_storage_type_t<T>* output,
+                                           data_type output_type)
+{
+  int const tid = threadIdx.x + blockIdx.x * blockDim.x;
+
+  auto const tdigest_index = tid / percentiles.size();
+  if (tdigest_index >= num_tdigests) { return; }
+  double const percentage = percentiles[tid % percentiles.size()];
+
+  // size of the digest we're querying
+  auto const tdigest_size = tdigest_offsets[tdigest_index + 1] - tdigest_offsets[tdigest_index];
+  if (tdigest_size == 0) {
+    // no work to do. values will be set to null
+    return;
+  }
+  double const* cumulative_weight = cumulative_weight_ + tdigest_offsets[tdigest_index];
+
+  // means and weights for the tdigest
+  double const* mean   = mean_ + tdigest_offsets[tdigest_index];
+  double const* weight = weight_ + tdigest_offsets[tdigest_index];
+
+  // min and max for the digest
+  double const* min_val = min_ + tdigest_index;
+  double const* max_val = max_ + tdigest_index;
+
+  double const total_weight = cumulative_weight[tdigest_size - 1];
+  double const cluster_q    = percentage * total_weight;
+
+  cast_output<T> cast_op{output_type};
+
+  if (cluster_q <= 1) {
+    output[tid] = cast_op(*min_val);
+    return;
+  } else if (cluster_q >= total_weight - 1) {
+    output[tid] = cast_op(*max_val);
+    return;
+  }
+
+  // otherwise find the centroid we're in and interpolate
+  size_type const centroid_index = static_cast<size_type>(
+    thrust::lower_bound(
+      thrust::seq, cumulative_weight, cumulative_weight + tdigest_size, cluster_q) -
+    cumulative_weight);
+
+  double diff = cluster_q + weight[centroid_index] / 2 - cumulative_weight[centroid_index];
+  if (weight[centroid_index] == 1 && std::abs(diff) < 0.5) {
+    output[tid] = cast_op(mean[centroid_index]);
+    return;
+  }
+  size_type left_index  = centroid_index;
+  size_type right_index = centroid_index;
+  if (diff > 0) {
+    if (right_index == tdigest_size - 1) {
+      output[tid] = cast_op(lerp(mean[right_index], *max_val, diff / (weight[right_index] / 2)));
+      return;
+    }
+    right_index++;
+  } else {
+    if (left_index == 0) {
+      output[tid] = cast_op(lerp(*min_val, mean[left_index], diff / (weight[left_index] / 2)));
+      return;
+    }
+    left_index--;
+    diff += weight[left_index] / 2 + weight[right_index] / 2;
```

Comment:
Can we add a comment on why the left_index side adds to `diff` but the right_index side doesn't?

Reply:
This function got a significant refactor.  Should be a lot more understandable (including comments on exactly what's going on with all the lerp stuff.